File: app/python/process.py
#!/usr/bin/python
#
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


"""
process.py - take data in raw_data bucket, process,
             and store in processed bucket.
"""

import csv
import json
import logging
import os

from config import (
    FACETS,
    PROCESSED_DATA_BUCKET,
    RAW_DATA_BUCKET,
    RAW_DATA_FILE,
    SEGMENTS,
)

import google.cloud.logging
import google.cloud.storage


# Enable Cloud Logging only when deployed to Cloud Run
if os.environ.get("K_SERVICE"):
    logging_client = google.cloud.storage.Client()
    logging_client.setup_logging()


def download_raw_data():
    """
    Download raw data from Cloud Storage into local file for processing
    """
    logging.info("  download_raw_data: start downloading data")

    if RAW_DATA_BUCKET is None:
        raise ValueError("RAW_DATA_BUCKET required")

    if PROCESSED_DATA_BUCKET is None:
        raise ValueError("PROCESSED_DATA_BUCKET required")

    temp_datafile = "test.csv"
    logging.info(
        "  download_raw_data: processing from "
        f"{RAW_DATA_BUCKET} to {PROCESSED_DATA_BUCKET}"
    )

    storage_client = google.cloud.storage.Client()
    raw_bucket = storage_client.get_bucket(RAW_DATA_BUCKET)

    raw_bucket.blob(RAW_DATA_FILE).download_to_filename(temp_datafile)

    logging.info("  download_raw_data: downloaded data to {temp_datafile}")
    return temp_datafile


def process_raw_data(temp_datafile):
    """
    Process local file, producing aggregate data
    """
    logging.info("  process_raw_data: start processing data")

    aggregate = {}
    ignored_records = 0
    counted_records = 0

    with open(temp_datafile) as f:
        csv_data = csv.DictReader(f)
        data = [row for row in csv_data]

        logging.info(f"Processing {len(data)} records from {RAW_DATA_FILE}")

    # Process each row.
    for row in data:

        # Ignore any records with incomplete data
        process = True
        for facet in FACETS:
            if row[facet] == "" or row[facet] == "?":
                ignored_records += 1
                process = False

        if process:
            # Build aggregate identifier
            row_key = "/".join([row[f] for f in FACETS])

            # Build the base data structure on first interaction
            if row_key not in aggregate.keys():
                aggregate[row_key] = {"_counter": 0}
                for segment in SEGMENTS:
                    if segment not in aggregate[row_key].keys():
                        aggregate[row_key][segment] = 0

            # Record the relevant data
            for segment in SEGMENTS:
                if row[segment] == "true":
                    aggregate[row_key][segment] += 1

            # Increment counters
            aggregate[row_key]["_counter"] += 1
            counted_records += 1

    logging.info(
        f"  process_raw_data: processed {counted_records} records,"
        f" removed {ignored_records}."
    )
    return aggregate


def write_processed_data(aggregate):
    """
    Write aggregate data to Cloud Storage
    """
    logging.info("  write_processed_data: start writing data.")
    counter = 0

    storage_client = google.cloud.storage.Client()
    processed_bucket = storage_client.get_bucket(PROCESSED_DATA_BUCKET)

    for rowkey in aggregate.keys():
        data_file = f"{rowkey}/data.json"
        facet_data = json.dumps(aggregate[rowkey])
        processed_bucket.blob(data_file).upload_from_string(facet_data)

        counter += 1

    logging.info(f"  write_processed_data: wrote {counter} files.")


if __name__ == "__main__":
    logging.info(
        f"🟢 Start process.py with: {RAW_DATA_BUCKET}, {PROCESSED_DATA_BUCKET}."
    )

    datafile = download_raw_data()
    result = process_raw_data(datafile)
    write_processed_data(result)

    logging.info("🏁 process.py complete.")

File: app/nodejs/process.js
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/*
process.js - take data in raw_data bucket, process,
             and store in processed bucket.
*/

'use strict';

import logger from './logger.js';
import fs from 'fs';
import os from 'os';
import path from 'path';

import csv from 'csv-parser';
import {Storage} from '@google-cloud/storage';

import config from './config.js';

const storage = new Storage();


/**
 * Download raw data from Cloud Storage into local file for processing
 * @return {string} Local filename containing downloaded data
 */
async function downloadRawData() {
  logger.info('downloadRawData: start downloading data');

  if (!config.RAW_DATA_BUCKET) {
    throw new Error('RAW_DATA_BUCKET required');
  }
  if (!config.PROCESSED_DATA_BUCKET) {
    throw new Error('PROCESSED_DATA_BUCKET required');
  }

  const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'rawData'));
  const tempDataFile = path.join(tempDir, 'raw_data.csv');

  logger.info(`downloadRawData: processing from ${config.RAW_DATA_BUCKET} ` +
      `to ${config.PROCESSED_DATA_BUCKET}`);

  await storage
      .bucket(config.RAW_DATA_BUCKET)
      .file(config.RAW_DATA_FILE)
      .download({destination: tempDataFile});

  logger.info(`downloadRawData: downloaded data to ${tempDataFile}`);

  return tempDataFile;
}
/**
 * Process local file, producing aggregate data
 * @param {string} tempDataFile Local filename containing downloaded data
 * @return {object} Correlated aggregate data
 */
async function processRawData(tempDataFile) {
  logger.info('processRawData: start processing data');

  return new Promise((resolve) => {
    const aggregate = {};
    let ignoredRecords = 0;
    let countedRecords = 0;
    const results = [];

    fs.createReadStream(tempDataFile)
        .pipe(csv())
        .on('data', (data) => results.push(data))
        .on('end', () => {
          results.forEach(function(row) {
            let validRow = true;

            config.FACETS.forEach(function(facet) {
              if ((row[facet] == '') || (row[facet] == '?')) {
                ignoredRecords += 1;
                validRow = false;
              }
            });

            if (validRow) {
            // build aggregate identifier
              const rowKeyValues = [];
              config.FACETS.forEach(function(facet) {
                rowKeyValues.push(row[facet]);
              });
              const rowKey = rowKeyValues.join('/');

              // Build the base data structure on first interaction
              if (!aggregate[rowKey]) {
                aggregate[rowKey] = {'_counter': 0};
                config.SEGMENTS.forEach(function(segment) {
                  if (!aggregate[rowKey][segment]) {
                    aggregate[rowKey][segment] = 0;
                  }
                });
              }

              // Record the relevant data
              config.SEGMENTS.forEach(function(segment) {
                if (row[segment] == 'true') {
                  aggregate[rowKey][segment] += 1;
                }
              });

              // Increment counters
              aggregate[rowKey]['_counter'] += 1;
              countedRecords += 1;
            }
          });

          logger.info(`processRawData: processed ${countedRecords} ` +
          `records, removed ${ignoredRecords}`);

          resolve(aggregate);
        });
  });
}


/**
 * Write aggregate data to Cloud Storage
 * @param {object} aggregate Aggregated data
 */
function writeProcessedData(aggregate) {
  logger.info('writeProcessedData: start writing');

  let counter = 0;

  const processedBucket = storage.bucket(config.PROCESSED_DATA_BUCKET);

  const writeData = new Promise((resolve) => {
    Object.keys(aggregate).forEach(async function(rowKey) {
      const facetData = aggregate[rowKey];
      processedBucket
          .file(`${rowKey}/data.json`)
          .save(JSON.stringify(facetData));
      counter += 1;
    });

    resolve(counter);
  });

  writeData.then(function(counter) {
    logger.info(`writeProcessedData: wrote ${counter} files`);
  });
}


const main = async () => {
  logger.info(`🟢 Start process.py with: ` +
    `${config.RAW_DATA_BUCKET},${config.PROCESSED_DATA_BUCKET}`);

  const dataFile = await downloadRawData();
  processRawData(dataFile)
      .then(function(result) {
        writeProcessedData(result);
      });
};

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
